### ISSUE TITLE
How do we re-initialize the GPU memory and states using cutorch? I find that we will kill the application when running cutorch, I often need to restart the computer or reinitialize it using some other tool. 

### ISSUE THREAD

**Issue by binuq8usa** [torch/cutorch#396]

**Comment by soumith** [comment]
the memory is automatically freed on the process exit. If you are seeing other behavior, then the process must be still running in the background.

**Comment by binuq8usa** [comment]
Thanks for the response. 
Well I did see the process and GPU utilization using "nvidia-smi" command from terminal. 

So, I tried killing the process and clearing the GPU memory from an external source. It was just hanging and so I kept restarting the computer to re-train a neural network with the necessary parameter changes. Is there some efficient way to cut short a training process (if training is not as expected), clear the GPU memory (and other variables), edit the parameters and re-train within the same/different instance of the itorch-notebook?

Note that I am running the network on a quadro card (mobile workstation) which has only 2GB of RAM.

**Comment by vince62s** [comment]
Hello,
I am having the same need. When freeing memory with (=nil) in cutorch, cutorch.getMemoryUsage() will report that the memory has been freed. however, cuda (nvidi-smi) still sees the memory as being used. So If we need another cutorch process to use some memory it may fail.
Is there any chance we can free the cuda memory without killing the lua process ?
cheers.
Vincent

**Comment by dno89** [comment]
Hello, same issue here.
Could it be related to the nvidia driver?

**Comment by pavanky** [comment]
@vince62s @dno89 I am assuming this is because of the caching memory manager that is currently enabled y default. Try launching with `THC_CACHING_ALLOCATOR=0` to have the memory released by the driver immedaitely.

**Comment by vince62s** [comment]
Indeed, the behavior is much better. After each Var = NIL and collectgarbage(), most of the memory is released, however, there is still a small amount being taken by the luajit process (about 300MB). Since all the alloc is done through a class.new, I would have thought the freeing of memory would be recursive. Not a big deal, but just wondering if there is some kind of overhead that could still be using a small portion.
